File: apps/buildsource/backend/src/routers/materials.py
"""Materials router."""

from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException, Query, status
from fastapi.security import HTTPBearer

from models.materials import (
    Material,
    MaterialCreate,
    MaterialUpdate,
    MaterialResponse,
    RegionalAvailability,
    SpecSheet,
    MaterialType,
)
from models.common import ApiResponse, PaginatedResponse
from services.baserow import get_baserow_service
from services.regional import get_regional_service

router = APIRouter(prefix="/materials", tags=["Materials"])
security = HTTPBearer()


@router.get("", response_model=ApiResponse[PaginatedResponse[MaterialResponse]])
async def list_materials(
    material_type: Optional[MaterialType] = None,
    search: Optional[str] = None,
    zip_code: Optional[str] = None,
    radius_miles: Optional[int] = Query(None, ge=0, le=500),
    min_recycled_content: Optional[float] = Query(None, ge=0, le=100),
    leed_eligible: Optional[bool] = None,
    in_stock: Optional[bool] = None,
    min_price: Optional[float] = None,
    max_price: Optional[float] = None,
    supplier_id: Optional[str] = None,
    astm_standard: Optional[str] = None,
    page: int = Query(1, ge=1),
    page_size: int = Query(24, ge=1, le=100),
    baserow=Depends(get_baserow_service),
) -> ApiResponse[PaginatedResponse[MaterialResponse]]:
    """List materials with filters."""
    filters = {}
    if material_type:
        filters["material_type"] = material_type.value
    if min_recycled_content is not None:
        filters["recycled_content_percent__gte"] = min_recycled_content
    if leed_eligible:
        filters["leed_points__gt"] = 0
    if supplier_id:
        filters["supplier_id"] = supplier_id
    if astm_standard:
        filters["astm_standard"] = astm_standard
    
    result = await baserow.list_rows(
        "products",
        filters=filters if filters else None,
        page=page,
        size=page_size,
    )
    
    materials = [MaterialResponse(**item) for item in result.get("results", [])]
    total = result.get("count", 0)
    
    return ApiResponse.success_response(
        PaginatedResponse.create(
            items=materials,
            total=total,
            page=page,
            page_size=page_size,
        )
    )


@router.get("/types", response_model=ApiResponse[List[str]])
async def get_material_types() -> ApiResponse[List[str]]:
    """Get all material types."""
    types = [t.value for t in MaterialType]
    return ApiResponse.success_response(types)


@router.get("/astm-standards", response_model=ApiResponse[List[str]])
async def get_astm_standards(
    baserow=Depends(get_baserow_service),
) -> ApiResponse[List[str]]:
    """Get all ASTM standards used in materials."""
    # Get unique ASTM standards from materials
    result = await baserow.list_rows("products", size=1000)
    standards = set()
    for item in result.get("results", []):
        if item.get("astm_standard"):
            standards.add(item.get("astm_standard"))
    
    return ApiResponse.success_response(sorted(list(standards)))


@router.get("/leed", response_model=ApiResponse[PaginatedResponse[MaterialResponse]])
async def get_leed_materials(
    page: int = Query(1, ge=1),
    page_size: int = Query(24, ge=1, le=100),
    baserow=Depends(get_baserow_service),
) -> ApiResponse[PaginatedResponse[MaterialResponse]]:
    """Get LEED-eligible materials."""
    result = await baserow.list_rows(
        "products",
        filters={"leed_points__gt": 0},
        page=page,
        size=page_size,
    )
    
    materials = [MaterialResponse(**item) for item in result.get("results", [])]
    total = result.get("count", 0)
    
    return ApiResponse.success_response(
        PaginatedResponse.create(
            items=materials,
            total=total,
            page=page,
            page_size=page_size,
        )
    )


@router.get("/search-by-spec", response_model=ApiResponse[List[MaterialResponse]])
async def search_by_spec(
    astm: str,
    grade: Optional[str] = None,
    baserow=Depends(get_baserow_service),
) -> ApiResponse[List[MaterialResponse]]:
    """Search materials by specification."""
    filters = {"astm_standard": astm}
    if grade:
        filters["grade_specification__contains"] = grade
    
    result = await baserow.list_rows("products", filters=filters, size=100)
    materials = [MaterialResponse(**item) for item in result.get("results", [])]
    
    return ApiResponse.success_response(materials)


@router.get("/{material_id}", response_model=ApiResponse[MaterialResponse])
async def get_material(
    material_id: str,
    baserow=Depends(get_baserow_service),
) -> ApiResponse[MaterialResponse]:
    """Get a single material by ID."""
    material = await baserow.get_row("products", material_id)
    if not material:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Material not found",
        )
    
    return ApiResponse.success_response(MaterialResponse(**material))


@router.get("/{material_id}/availability", response_model=ApiResponse[List[RegionalAvailability]])
async def get_regional_availability(
    material_id: str,
    zip_code: str,
    radius: int = Query(50, ge=0, le=500),
    baserow=Depends(get_baserow_service),
    regional=Depends(get_regional_service),
) -> ApiResponse[List[RegionalAvailability]]:
    """Get regional availability for a material."""
    # Get availability data
    result = await baserow.list_rows(
        "regional_availability",
        filters={"material_id": material_id},
    )
    
    availabilities = []
    for item in result.get("results", []):
        # Calculate distance
        distance = 0  # Would calculate actual distance
        if distance <= radius:
            availabilities.append(RegionalAvailability(**item))
    
    return ApiResponse.success_response(availabilities)


@router.get("/{material_id}/spec-sheets", response_model=ApiResponse[List[SpecSheet]])
async def get_spec_sheets(
    material_id: str,
    baserow=Depends(get_baserow_service),
) -> ApiResponse[List[SpecSheet]]:
    """Get spec sheets for a material."""
    result = await baserow.list_rows(
        "spec_sheets",
        filters={"material_id": material_id},
    )
    
    sheets = [SpecSheet(**item) for item in result.get("results", [])]
    return ApiResponse.success_response(sheets)


@router.post("/compare", response_model=ApiResponse[List[MaterialResponse]])
async def compare_materials(
    material_ids: List[str],
    baserow=Depends(get_baserow_service),
) -> ApiResponse[List[MaterialResponse]]:
    """Compare multiple materials."""
    materials = []
    for material_id in material_ids:
        material = await baserow.get_row("products", material_id)
        if material:
            materials.append(MaterialResponse(**material))
    
    return ApiResponse.success_response(materials)


@router.get("/{material_id}/related", response_model=ApiResponse[List[MaterialResponse]])
async def get_related_materials(
    material_id: str,
    limit: int = Query(5, ge=1, le=20),
    baserow=Depends(get_baserow_service),
) -> ApiResponse[List[MaterialResponse]]:
    """Get related materials."""
    # Get the material first
    material = await baserow.get_row("products", material_id)
    if not material:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Material not found",
        )
    
    # Find related materials (same type, different supplier)
    result = await baserow.list_rows(
        "products",
        filters={
            "material_type": material.get("material_type"),
        },
        size=limit + 1,  # +1 to account for excluding self
    )
    
    related = []
    for item in result.get("results", []):
        if str(item.get("id")) != material_id:
            related.append(MaterialResponse(**item))
        if len(related) >= limit:
            break
    
    return ApiResponse.success_response(related)

File: apps/buildsource/frontend/src/services/materials.ts
import { api } from './api';
import { 
  Material, 
  MaterialFilters, 
  PaginatedResponse, 
  PaginationParams,
  RegionalAvailability,
  SpecSheet 
} from '@types/index';

export const materialsService = {
  // Get all materials with filters
  getMaterials: async (
    filters?: MaterialFilters,
    pagination?: PaginationParams
  ): Promise<PaginatedResponse<Material>> => {
    const params = new URLSearchParams();
    
    if (filters) {
      if (filters.material_type) params.append('material_type', filters.material_type);
      if (filters.search) params.append('search', filters.search);
      if (filters.zip_code) params.append('zip_code', filters.zip_code);
      if (filters.radius_miles) params.append('radius_miles', filters.radius_miles.toString());
      if (filters.min_recycled_content) params.append('min_recycled_content', filters.min_recycled_content.toString());
      if (filters.leed_eligible) params.append('leed_eligible', 'true');
      if (filters.in_stock) params.append('in_stock', 'true');
      if (filters.min_price) params.append('min_price', filters.min_price.toString());
      if (filters.max_price) params.append('max_price', filters.max_price.toString());
      if (filters.supplier_id) params.append('supplier_id', filters.supplier_id);
      if (filters.astm_standard) params.append('astm_standard', filters.astm_standard);
    }
    
    if (pagination) {
      params.append('page', pagination.page.toString());
      params.append('page_size', pagination.page_size.toString());
    }
    
    const response = await api.get<PaginatedResponse<Material>>(`/materials?${params.toString()}`);
    return response.data!;
  },

  // Get single material
  getMaterial: async (id: string): Promise<Material> => {
    const response = await api.get<Material>(`/materials/${id}`);
    return response.data!;
  },

  // Get material types
  getMaterialTypes: async (): Promise<string[]> => {
    const response = await api.get<string[]>('/materials/types');
    return response.data!;
  },

  // Get ASTM standards
  getASTMStandards: async (): Promise<string[]> => {
    const response = await api.get<string[]>('/materials/astm-standards');
    return response.data!;
  },

  // Get regional availability for a material
  getRegionalAvailability: async (
    materialId: string,
    zipCode: string,
    radiusMiles: number = 50
  ): Promise<RegionalAvailability[]> => {
    const response = await api.get<RegionalAvailability[]>(
      `/materials/${materialId}/availability?zip_code=${zipCode}&radius=${radiusMiles}`
    );
    return response.data!;
  },

  // Get spec sheets for a material
  getSpecSheets: async (materialId: string): Promise<SpecSheet[]> => {
    const response = await api.get<SpecSheet[]>(`/materials/${materialId}/spec-sheets`);
    return response.data!;
  },

  // Compare materials
  compareMaterials: async (materialIds: string[]): Promise<Material[]> => {
    const response = await api.post<Material[]>('/materials/compare', { material_ids: materialIds });
    return response.data!;
  },

  // Get related materials
  getRelatedMaterials: async (materialId: string, limit: number = 5): Promise<Material[]> => {
    const response = await api.get<Material[]>(`/materials/${materialId}/related?limit=${limit}`);
    return response.data!;
  },

  // Get LEED-eligible materials
  getLEEDMaterials: async (
    pagination?: PaginationParams
  ): Promise<PaginatedResponse<Material>> => {
    const params = new URLSearchParams();
    if (pagination) {
      params.append('page', pagination.page.toString());
      params.append('page_size', pagination.page_size.toString());
    }
    const response = await api.get<PaginatedResponse<Material>>(`/materials/leed?${params.toString()}`);
    return response.data!;
  },

  // Search materials by specification
  searchBySpec: async (
    astmStandard: string,
    grade?: string
  ): Promise<Material[]> => {
    const params = new URLSearchParams();
    params.append('astm', astmStandard);
    if (grade) params.append('grade', grade);
    const response = await api.get<Material[]>(`/materials/search-by-spec?${params.toString()}`);
    return response.data!;
  },
};
